algebird-core/src/main/scala/com/twitter/algebird/Window.scala

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 Stripe
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird
+
+import scala.collection.immutable.Queue
+
+/**
+ *
+ * Convenience case class defined with a monoid for aggregating elements over
+ * a finite window.
+ *
+ * @param total wrapped instance of `T`
+ * @param q (optional) queue to override default q containing input "total"
+ */
+
+case class Window[T](total: T, q: Queue[T] = Queue()) {
```

Comment:
This looks great! What do you think about allowing user to pass a `PriorityQueue` instead of just a Queue (AFAIK scala PQ doesnt extend Queue). The way code is written right now, its implying that I need a commutative monoid to guarantee deterministic behavior or I need to control the order in  which items are inserted in the window. When I work with windows they usually have  some sort of ordering defined (last 3 distinct IPs used by a user in last one hour for example), It would be nice to allow users to  preserve the ordering somehow IMO

Reply:
actually, I think `Window[T](total: T, items: Queue[T])` is better. We don't really want people to do something like `Window(100)` do we? that will have no items in the queue, and I imagine break the laws.

Reply:
If you look at `items`, it fills in the queue if none was provided, i.e meant for initialization. The intent is to have a simple interface where you can wrap a window around a single object and sum in a sensible way:
```
implicit wm = WindowMonoid(2)

Window(1) + Window(2) + Window(3) == Window(5)
```
Needing to specify the queue would make this redundant 
```
Window(1, Queue(1)) + ...
```

Tbh I'm not a big fan of even having to expose the queue, since you could change the implementation in the future (e.g. a PriorityQueue). A more general approach might use a `PushPopper` interface, but that's probably overkill.